New module SupplyChain

aleeciu opened the pull request:
This is an substatial update and rethinking of the supplychain module. Previosuly the module was running an Input-Output calculation for the whole world with pre-specified exposures. Now it can deal with user-specified areas and exposures. I also added a functionality to download the Input-Output table for the year of interest and added a tutorial which will be improved in the future.

Remaining issues (both for @emanuel-schmid):
1. I put the test data, as usual, into climada_python\climada\engine\test\data; however with the new CONFIG settings I cannot access this folder as when running the test functions it by deafult goes into ~/climada/data/ where there are no test data... I do not know how to deal with this.. (moving them manually isn't an option I guess)
2. To download the files we will need pyxlsb (which I also added in the new env_climada.xml), is that an issue?

chahank reviewed (changes requested):
Thank you very much for rewriting the previous supplychain module! It is now in a very good state. 

Below are nevertheless quite a few change suggestions. In particular, I think the handling of the data and how to address the data matrix could be improved. The review might come out as a bit confusing. I had to stop before finishing, as I think the data matrix handling should be discussed (addressed) first.

aleeciu commented on climada/engine/__init__.py:
```diff
@@ -20,3 +20,4 @@
 """
 from .impact import *
 from .cost_benefit import *
+from .supplychain import *
```
What do you mean by submodule of engine? Isn't it what it currently is? Anyway I just left were it was..

chahank replied:
Make a folder called `supplychain` and add the file `supplychain.py` in there.

aleeciu replied:
Mh I see but supplychain is only one file.. shall we need a submodule for one func (for unceratinty, for instance, there were more files)? Anyway it is fine for me, I think it is more a decision on when we want to make new submodules in climada and when not.

chahank replied:
Let me add @emanuel-schmid to the conversation. I think it is cleaner to make submodules, even if they have only a single file. Because `supplychain` or `uncertainty`  are simply not on the same level as `impact` or `costbenefit`. @emanuel-schmid , what do you think?

emanuel-schmid replied:
As long as there is only one file belonging to this module, `climada/engine/supplychain.py` should be fine I guess.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+from climada.entity.exposures.base import Exposures
+
+LOGGER = logging.getLogger(__name__)
+WIOD_FILE_LINK = 'http://www.wiod.org/protected3/data16/wiot_ROW/'
```
Mh no, I was not part of that conversation. I can move that if needed.

chahank replied:
Well, then since we do not know for sure, let's leave it here.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+        """Calculate for each country/sector-combination the direct impact per year.
+        I.e. compute one year impact set for each country/sector combination. Returns
+        the notion of a supplychain year impact set, which is a dataframe with size
+        (n years) * ((n countries)*(n sectors)).
```
The shape needs to be the same for the calcualtion below and there will be zeros in sectors not included in sector_type..

chahank replied:
Ok. Can this be included in the docstring please?

aleeciu replied:
as above

chahank replied:
??

aleeciu replied:
i.e., as the user must know what the WIOD table structure is, I find info regarding the table structure (what is on cols, what on rows, table dimensions etc. etc.) redundant and unnecessary. Also, the module comes with a tutorial which also touches upon this. These detailed comments are residuals from the previous implementation.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+
+        sectors = mriot.iloc[row_st:row_sect_end, col_sect].values
+        countries_iso3 = np.unique(mriot.iloc[row_st:row_end, col_iso3])
+        # move Rest Of World (ROW) at the end of the array as countries
```
ROW is a "country" of the WIOT table - it is assumed whoever uses the module knows what the WIOT table is and how it is structured otherwise it is hard to know what the module does..

chahank replied:
aaa ok, I misundertood that, because below in the code it says that the `ROW` can be multiples that need to be aggregated

>           # Sum needed below in case of many ROWs, which need be aggregated.
            # Another option is to keep them separate, but it requires changing
            # all functions below as impact matrix will no longer have
            # dim = (n_years, self.n_countries*self.n_sectors) but
            # dim = (n_years, (self.n_countries+#ROWs_countries-1)*self.n_sectors)

aleeciu replied:
Yes. So basically you have these bunch of countries from the WIOT table, plus one which is ROW. Then, for to the countries you have exposure and impact for, you must check whether this is explicitely included in the WIOT or it falls in ROW..

chahank replied:
I see. It would probably be simpler to compute the impact for all the `ROW` at once then no?

aleeciu replied:
You need to iterate to known which one is ROW and which one is not - in the meantime you can assess their damage

chahank replied:
Why? You can select all the exposures at once that are in the `ROW`, and then compute the impact from that...

aleeciu replied:
Ok, so what happens here is that all countries which are NOT explicitely modeled by the WIOD table ends up in ROW. Therefore, in case one is dealing with more than a country which ends up in ROW, their overall impact is summed up so to give the ROW "country" overall impact. This guarantees consistency with the dimensions of the WIOD table. In general, I do not think it makes sense modelling _any_ country which ends up in ROW, as that is a very crude approximation of their IO structure. The comment below has been reduced as I found it myself quite consusing (or, to the least, providing unnecessary info)

```
      # Sum needed below in case of many ROWs, which need be aggregated.
       # Another option is to keep them separate, but it requires changing
       # all functions below as impact matrix will no longer have 
      # dim = (n_years, self.n_countries*self.n_sectors) but
       # dim = (n_years, (self.n_countries+#ROWs_countries-1)*self.n_sectors)
```

chahank commented:
Yes, it is included in the comments. I think it is easiest if you first have a look at the comments. In summary, I do not understand why one would bother with a complex index juggling instead of using the capacities of dataframes to select columns easily.

aleeciu commented on requirements/env_climada.yml:
```diff
@@ -23,6 +23,7 @@ dependencies:
   - pyepsg>=0.4
   - pytables>=3.6
   - python=3.8
+  - pyxlsb=1.0.8  
```
Yes I actually asked emanuel in the merge request about this. It is needed if we want to download the WIOT tables, otherwise we would need to store them in CLIMADA and they are big exceel sheets (something about 2400x2400 matrixes).

chahank replied:
Perfect, thanks for checking with Emanuel!

aleeciu replied:
Well let's see what he thinks..

aleeciu replied:
It seems it can be added

aleeciu commented:
> Yes, it is included in the comments, but not so explicit, as I started to get confused. In summary, I do not understand why one would bother with a complex index juggling instead of using the capacities of dataframes to select columns easily.

If you are referring to the functions in the indirect impact function, I agree they could be written in a better way, but I did not touch them from the previous implementation. Those blocks were supposed to rightly model the various IO models and we agreed to review the package in order to make it working and not to rewrite it completely.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+    Attributes:
+        mriot_data (np.array): 2-dim np.array of floats representing the data of
+            a full multi-regional input-output table (mriot).
+        countries (np.array): 1-dim np.array of strings containing the full list
```
countries are spelled out, countries_iso3 respect the same order and provide iso3 codes

chahank replied:
Spelled out according to what convention? ISO3?

aleeciu replied:
countries have now been removed since we have countries_iso3 - that is how both CLIMADA and WIOD work

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+        countries_iso3 (np.array): similar to .countries, but containing the
+            countries' respective iso3-codes.
+        sectors (np.array): 1-dim np.array of strings containing the full
+            list of sectors represented in the mriot, corresponding to the columns/
```
Both countries and sectors are in both rows/columns. 

Columns and rows of the WIOT table have countries, and for each country, values for each sector. So the num of cols/rows = n_countries x n_sectors

aleeciu replied:
That info is now been removed all together as whoever uses this package must know how the WIOD table is structured

chahank replied:
In what commit? Please refer to the changes so that one can see what was done. I am not yet convinced by your argument here.

aleeciu replied:
the changes were made in commit 4d04eaca49dd93bc711f487be24c2b6dabe8f9c5

If you are up to use this module, which uses WIOD IO tables at its core, I think we can agree one needs to know what this tables are, and how they are structured; then it follows one knows what is on the cols, what is on the rows therefore there is not need to repeat it in the docstring.

chahank replied:
That is a fair point. However, it makes it rather difficult to review and/or debug in the future, in particular the parts where indices cuts are made in the code which are not obvious. But, you can leave it as is, I guess it is a matter of taste. Thanks for linking to the commit!

aleeciu replied:
The only hardcoded bit are lines 131-135:
```
        col_sectors = 1
        col_iso3 = 2
        end_row_sectors = 61
        start_row, end_row = (5, 2469)
        start_col, end_col = (4, 2468)
```
I think for debugging one must know what is the data structure he/she is dealing with. There are also info in the tutorial (which could be improved indeed) regarding data structure etc...

chahank replied:
Sorry for the confusion, the hard-coded indices have nothing to do with that part, you are correct. My mistake. 
However, saying what the WIOT table is made of is useful I think, since it is not necessary to use WIOD tables in this module, but any WIOT table is possible.

chahank commented:
I do not think that we speaking of the same point. My comment is about the handling of the wiot and extracting the sectoral values from it including subsection. For example in lines 

```
            values_pos_cntry = range(step_in_table+init_pos,
                                     step_in_table+end_pos)
            subsec_cntry_prod = self.mriot_data[values_pos_cntry].sum(axis=1)
```

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+        file_name = 'WIOT{}_Nov16_ROW.xlsb'.format(year)
+        results_folder = Path(file_path).joinpath('results')
+
+        if not user_data:
```
As for the `user_data`, I am thinking of a way to drop it.

As for renaiming `results`, that is the name the new CONFIG gives to whatever is downloaded, how can I change that?

chahank replied:
Really? Ok, then we probably should not change it just for this pull request. Thanks for the reminder. Let us take this up with @emanuel-schmid .

aleeciu replied:
No need for  `user_data` anymore.. as per the second point now by default it goes into a 'result' directory (as per CONFIG) but the user can specify whatever folder he wants

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            mriot = pd.read_excel(os.path.join(file_path, file_name),
+                                  engine='pyxlsb')
+
+        row_st,row_end=rows_range
+        col_sect,row_sect_end=cols_sect_range
+        col_data_st,col_data_end=cols_data_range
```
I am missing what is called `_dt`

chahank replied:
Sorry, typo `_st`. Second line is `col_sect` and first/third have `_st`

aleeciu replied:
`_st` stands for start

aleeciu replied:
anyway changed now into `start`

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+        self.n_countries = n_countries
+        self.n_sectors = n_sectors
+        self.cntry_pos = cntry_pos
+        self.mriot_type = 'wiod'
```
Informational, but currently there are no other choices than using World Input-Output Tables (WIOT) from the [World Input-Output Database (WIOD) project](http://www.wiod.org/home) so it is not that informative. If more tables will be added then we will need another read_table function and the mriot_type could become more informative.

chahank replied:
Good idea to include this. Please make this clear in the docstring.

aleeciu replied:
There is, along with references.

chahank replied:
What commit? Also, please indicate hwat references you mean. Finally, the docstring  Also, the docstring is not understandable currently (and I think incorrect):
```
    mriot_type : str
        Type of the adopted input-output table. Currently only WIOD tables
        are supported.
```
The code support other tables than 'WIOD'. It is just that only for WIOD tables there is automatic data retrieval. Furthermore, `Type of the adopted input-output table` furthermore is unclear. Does it mean "sources of the data"?

aleeciu replied:
That is not correct. The code is hardcoded for WIOD tables only. This is made clear since the very name of the function (read_wiod16 - there is no other function which reads IO tables), as well as repeated in the docstring description. References (bibliography) are even provided in the function description.

chahank replied:
But I do not understand why `calc_sector_direct_impac` and `calc_sector_indirect_impac` would not work for any other WIOT table ... or rather, these methods should work irrespective of the data source, as long as the data is loaded in the same way as does `read_wiot16`.

aleeciu replied:
Those functions will indeed work - but currently the only type of IO we read are WIOD; thus the quote

> Currently only WIOD tables are supported.

even if `calc_sector_direct_impac` and `calc_sector_indirect_impac `funcs are general, you can currently only use them for WIOD tables

chahank replied:
That is incorrect formulation. The class works very well even without WIOD tables. In this case one however must supply that data for `self.sectors`, `self.countries_iso3`, `self.mriot_data`, `self.total_prod`, `self.cntry_pos`  and `self.mriot_type`.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+        self.n_sectors = 0
+        self.mriot_type = 'None'
+
+    def read_wiot(self, year=2014, file_path=SYSTEM_DIR, user_data=False,
```
mh no, maybe read_wiod - basically the WIOT of the WIOD project; if more will come they will be read_whatever

chahank replied:
so then `read_wiod` ?

aleeciu replied:
yes

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            hazard (Hazard): Hazard object for impact calculation.
+            exposure (Exposures): Exposures object for impact calculation.
+            imp_fun_set (ImpactFuncSet): Set of impact functions.
+            sec_subsec (list): User-defined list with the start and end positions
```
Dealing with row and column names is not possible. We are talking about over 2400 rows/columns, where names in fact replicate. For n sectors, you haven n rows/cols for country 1, then n rows/cols for country 2,...., n rows/cols for country m.. and they are named the same. The easiest way to access it is to know at which pos is your country, and then which of the n sectors you need. Which also points to your comment below about this piece of code:

`            values_pos_cntry = range(step_in_table+init_pos,
                                     step_in_table+end_pos)
            subsec_cntry_prod = self.mriot_data[values_pos_cntry].sum(axis=1)
`

this is explained in the tutorial

chahank replied:
Yes. But why then not use the already existing separator `self.cntry_pos` instead of building them new every time? Also, why define the `subsec` as neigbours in the table?

Here I would expect that you use two functions to select
1. `self.cntry_pos` to select by country
2. subsec_pos to select the position of the sections inside of the country rows/columns. This should simply be a list of the positions instead of the start/end positions, as this restricts the use without clear reason.

This way the code becomes clearer and it is more robust to errors in indexing.

aleeciu replied:
The code has been updated.. now I use `self.cntry_pos` and allow non continuous indexing

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            rows_range (tuple): Start and end position of rows with data
+            col_iso3 (int): pos of col with iso3 countries
+            cols_data_range (tuple): start and end position of cols with data
+            cols_sect_range (tuple): start and end position of sectors' names
```
since these kwargs are hardcoded and I see no reason to change them since the function is meant to be used to solely work with WIOD tables (fixed structure), I removed them and keep fix values within the function

chahank replied:
Either these are kwargs with default values that could be set to another value for a given reason or they are unchangeable 'hardcoded' and then are not kwargs. 

In any case, the default for kwargs must be documented in the docstrings.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            try:
+                countries.append(countries_by_alpha3[iso3][0])
+            except KeyError:
+                countries.append('Rest of World')
```
done

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+        years = np.unique([date.year for date in dates])
+
+        # Keep original order of countries
+        _, cntry_idx = np.unique(exposure.gdf.region_id, return_index=True)
+        unique_regid_same_order = exposure.gdf.region_id[np.sort(cntry_idx)]
```
applied

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+                                        self.n_countries*self.n_sectors))
+
+        cntry_dir_imp = []
+        for cntry in unique_regid_same_order:
```
The WIOD table works with countries... if one wants to model an admin1 territory he must pass it with the iso3 code of the respective country. Using this module requires knowing the structure of the WIOD table...

chahank replied:
Exactly my point. If I have a WIOT for admin1 territory from some source, how can I use this code which requires iso3 ids?

aleeciu replied:
I don't understand what is it that I should make clearer. If you have a IO table at the admin1 level, you should first write your own function to read such table and then I see no reason of using anything else than an admi1 exposure.

chahank replied:
Why would one need another function to read the table? My understanding is that the table consists of columns and rows that have two types of subdivision, one that macro division into geographical regions (for WIOD countries), and one sub-division for different sectors. This is what you function is doing, taking the inputs how to cut among these two divisions. The rest of the impact computation is completely agnostic to what the data represents (whether countires or oder regions).

aleeciu replied:
This *one* table is structured that way, and I only know this one - in fact I inherited this module and this was the only contemplated table. I don't know how other tables might be structured. Given this, how can one claim any sort of generalizability? I am more comfortable in leaving this as a `read_wiod` (and *only* wiod function), the same we have `read_ibtracs`, `read_emanuel`, `read_whatever` functions for tropical cyclones.

chahank replied:
I understand you point. Note that your example with `read_ibtracs` is actually exactly supporting my argument, not yours. The TCTracks module has one function to read data sources `read_ibtracs`. However, the rest of the module does not care whether the data comes from `IBTrACS` or any other source, as long as the data is read in the class correctly. I suggest you to write the code the same way. As far as I understood, this is actually already mostly the case.

aleeciu replied:
But that is the case, the rest of the code works for other tables provided that the are loaded correctly (same as TC with different data sources). But currently only one data source if supported: WIOD table.

chahank replied:
A module in CLIMADA should always separate computations from data loading. Your module does that very well (up to one minor point I think the `calc_direct_impact` method which I will point out below). It has methods that do supply chains. Then, there is for convenience one method to load data. The Docstring should reflect this. Any user can supply their own WIOT table and use this module. It is absolutely NOT limited to WIOD, but it only allows to automatically set the data for WIOD.

aleeciu replied:
Honestly I do not know as I have never dealt with any other IO table. I simply cannot be sure, though it is possible, that other tables follow the same structure as WIOD and can thus be loaded by simply adjusting params of the existing fun, which is btw called `read_wiod16`

chahank replied:
It actually does not matter much whether other IO tables are the same or not. What matters is that you defined a data structure (the one from WIOD), and any data inputed in that structure works. It is important to separate data acquisition and data computations.

aleeciu replied:
In https://github.com/CLIMADA-project/climada_python/pull/180/commits/ed468290bf322d343c230549529e51c09832e9d0 the code is made general now by renaming countries in more general regions and by providing a mapping function in the direct damage calculation.

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            Default values of the last four args allow reading the full wiot table.
+        """
+
+        file_name = 'WIOT{}_Nov16_ROW.xlsb'.format(year)
```
I left it how it was unless there is a strong reason to do otherwise

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            sector_type (str): If sec_subsec is not defined; it sets the start
+                and end positions in the mriot tablefor some default sectors.
+                Possible values are "service", "manufacturing", "agriculture"
+                and "mining". Either sec_subsec or sector_type must be defined.
```
`sector_type` is now set to default value `service`

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+            idx_country = np.where(self.countries_iso3 == cntry_iso3)[0]
+
+            if idx_country.size > 0.:
+                step_in_table = idx_country[0]*self.n_sectors
```
removed and coded updated

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,316 @@
+            hazard (Hazard): Hazard object for impact calculation.
+            exposure (Exposures): Exposures object for impact calculation.
+            imp_fun_set (ImpactFuncSet): Set of impact functions.
+            selected_subsec (list): User-defined list with positions of the
+                                    subsector to analyze. Default None.
+            sector_type (str): If sec_subsec is not defined; it sets the start
+                and end positions in the mriot tablefor some default sectors.
+                Possible values are "service", "manufacturing", "agriculture"
+                and "mining". Either sec_subsec or sector_type must be defined.
```
Question: what happens when both `sector_type` and `selected_subsec` are set?

Suggestion: make only one variable which is either a list with positions or a string. although it is slightly transvestite in types, this is not unseen in Python and could simplify the whole thing. 
```suggestion
            selected_subsec : list(int) or string
                Selection of sectors/subsectors to aggregate. It is either 1) a list of the index positions of the
                selected sectors/subsectors in the WIOT table, or 2) one of the pre-defined list of subsectors
                "service", "manufacturing", "agriculture", or "mining" for the WIOD data.
                 The default is "service".
```

aleeciu replied:
done

chahank commented on climada/engine/test/test_supplychain.py:
```diff
@@ -0,0 +1,180 @@
+        impf_set.check()
+        
+        # Test direct impacts
+        sup.calc_sector_direct_impact(hazard, exp, impf_set,
```
PLease add a test with `set_subsec = [...]`

aleeciu replied:
done in https://github.com/CLIMADA-project/climada_python/pull/180/commits/ed468290bf322d343c230549529e51c09832e9d0

chahank commented on climada/engine/test/test_supplychain.py:
```diff
@@ -0,0 +1,180 @@
+        self.assertAlmostEqual(sup.total_prod.sum(), 3533367.89439, places=3)
+        self.assertEqual(len(sup.countries_iso3), len(sup.countries))
+
+    def test_calc_impact(self):
```
I am not sure about this, but I would think it is easier to debug if this function is separated into several functions which each test one single function of the code. @emanuel-schmid , could you please comment?

emanuel-schmid replied:
Do you mean to separate `sup.calc_sector_direct_impact` from `sup.calc_indirect_impact`? Then I agree, it's best to write unit tests around exactly one method/function to be scrutinized.

aleeciu replied:
Yes but in oder to run  `sup.calc_indirect_impact` you need to have previosuly run `sup.calc_sector_direct_impact` that is why I put them in the same test

chahank replied:
I still would do separate test functions, otherwise, if it fails you do not know whether it is `sup.calc_indirect_impact` or `sup.calc_sector_direct_impact` that has a problem.

chahank commented:
Thanks for the many updates! One big issue has not been addressed: the docstrings must be put in NumpyDoc format. C.f. the developer's guides.

emanuel-schmid commented on data/desktop.ini:
```diff
@@ -0,0 +1,2 @@
+[LocalizedFileNames]
+ECLIPSE_base_CLE_V5a_NOx.nc=@ECLIPSE_base_CLE_V5a_NOx.nc,0
```
What is the purpose of this file?

aleeciu replied:
I have no idea what that is... I just noticed that, it was created automatically for reasons I ignore..

aleeciu replied:
removed

aleeciu commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,395 @@
+    """SupplyChain definition. It provides methods for the entire supplychain-risk
+    workflow and attributes holding the workflow's data and results.
+
+    Attributes:
```
done

emanuel-schmid commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,316 @@
+from climada.entity.exposures.base import Exposures
+
+LOGGER = logging.getLogger(__name__)
+WIOD_FILE_LINK = 'http://www.wiod.org/protected3/data16/wiot_ROW/'
```
The url should go into the config file.

aleeciu replied:
done in https://github.com/CLIMADA-project/climada_python/pull/180/commits/ed468290bf322d343c230549529e51c09832e9d0

emanuel-schmid commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,316 @@
+        E., Los, B., Stehrer, R. and de Vries, G. J. (2015), "An Illustrated
+        User Guide to the World Input–Output Database: the Case of Global
+        Automotive Production", Review of International Economics., 23: 575–605
+
```
Since the method is to a large extent hard coding and dealing with the Release 2016 exclusively, I'd suggest to change the name into `read_wiod_2016` or something.

aleeciu replied:
mhh I find that a bit misleading as one could think is the WIOD table of year 2016, maybe something like wiod16

aleeciu replied:
updated

aleeciu commented:
I have now also updated the docstring. I think I addressed all comments now. The only remaining issues are:

1. As also stated in the pull request's main comment above, I do not know how to deal with the test file data after the new CONFIG settings. Specifically, I have two issues:

     1.1 There are some new data to be used in test which are not in the SYSTEM_DIR folder (user_name/climada) which is by deafult where the system expects the data to be. The test data are instead in the CLIMADA repository folder (climada_python\climada\engine\test\data), I could specify this manually but it would crash when other people will run it.
 
     1.2 I do not know how to access test data already in the climada repository. Following what is e.g. in 
climada_python\climada\engine\test\test_impact.py; I wrote something like:

```
DATA_FOLDER = CONFIG.engine.test_data.dir()
HAZ_TEST_MAT = CONFIG.hazard.test_data.dir().joinpath('atl_prob_no_name.mat')
```
but the first crashed as there is no `test_data`, the latter as there is no `engine`. This also means that when I run the exisitng (and supposedly working) `test_impact`, then it crashes.

2. make new tests as suggested (easy after the above is fixed)
3. add 
`WIOD_FILE_LINK = 'http://www.wiod.org/protected3/data16/wiot_ROW/'`

in CONFIG (as per @emanuel-schmid comment) but don't know how to do it.

emanuel-schmid commented:
1. But there is. When you run the tests from `climada_python\` the config file `climada_python\climada.conf` ist taken into account too. And there the test data folder is already configured, pointing to `climada_python\climada\engine\test\data`.
When you run the test from another working directory, copy over the config file.

3. add an entry in the config file:
```
    "engine": {
        "supplychain": {
            "resources": {
                "wiod_16": "http://www.wiod.org/protected3/data16/wiot_ROW/"
            }
        }
    },
```
and refer to it like so:
```
WIOD_FILE_LINK = CONFIG.engine.supplychain.resources.wiod_16
```

aleeciu commented:
ERRATA CORRIGE to point 1.2 above:

however, when running the whole test_file I get:
```
HAZ_TEST_MAT = /climada_python/climada/engine/test/climada/hazard/test/data/atl_prob_no_name.mat'
DATA_FOLDER = /climada_python/climada/engine/test/climada/engine/test/data'
```
instead of:
```
HAZ_TEST_MAT = /climada_python/climada/hazard/test/data/atl_prob_no_name.mat'
DATA_FOLDER = /climada_python/climada/engine/test/data'
```

so it is building the paths on the path `/climada_python/climada/engine/test` (i.e., the current directory as my running test_impact.py) instead on the path `/climada_python/`

so at the end it crashes..

emanuel-schmid commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,322 @@
+        self.total_aai_agg = np.array([], dtype='f')
+        self.io_data = {}
+
+    def read_wiod16(self, year=2014, file_folder=SYSTEM_DIR.joinpath('results')):
```
Although it's probably fine, function calls in optional parameter default assignment make me uneasy. I'd prefer `..., file_folder=None):` and then in the method block:
``` if file_folder = file_folder or SYSTEM_DIR.joinpath('results')```

Is `climada/system/results` really the preferred path for WIOD tables, which are basically input data?

Actually - wouldn't it be appropriate to have a folder `climada/system/wiod` just fort this? In this case you can also remove the parameter altogether: wiod data goes there. period.
You could use the config file to specify it:
"engine": {
        "supplychain": {
            "resources": {
                "wiod_16": "http://www.wiod.org/protected3/data16/wiot_ROW/"
            },
            "local_data": "{local_data.system}/WIOD"
        }
    },

Only need to make sure the test file is found during test. Either copy it over there or, even better, overwrite the config values.

aleeciu replied:
Thanks @emanuel-schmid. This is exactly the kind of tip I was looking for. I am only missing how to do the last, i.e., make sure that the test files are found under `CONFIG.engine.supplychain.resources.wiod_16.str()`

ps
the test files **cannot** be downloaded

aleeciu replied:
Suggestions were well received and implemented, thanks. I am just unsure whether I am doing the below effectively:

> Only need to make sure the test file is found during test. Either copy it over there or, even better, overwrite the config values.

as I copy the file from the test folder to the system/WIOD folder using shutil.copy (see line 42 of test_supplychain.py in commit https://github.com/CLIMADA-project/climada_python/pull/180/commits/ed468290bf322d343c230549529e51c09832e9d0)

other/better ways are welcome; please explicit code in case.

emanuel-schmid commented:
> so it is building the paths on the path `/climada_python/climada/engine/test` (i.e., the current directory as I am running test_impact.py) instead on the path `/climada_python/`
>

Well, then change the current directory. You can run test_impact.py from anywhere - can't you?

aleeciu commented:
> Well, then change the current directory. You can run test_impact.py from anywhere - can't you?

sure, but shoudn't this work automatically with the pre-defined settings? I mean if one doanwloads the package and then run any test, this should not crash..

emanuel-schmid commented:
If they _must_ run the tests they can run it from `climada_python`, same as Jenkins does. If the test passes for Jenkins, it will pass for them too.

emanuel-schmid commented on requirements/env_climada.yml:
```diff
@@ -23,6 +23,7 @@ dependencies:
   - pyepsg>=0.4
   - pytables>=3.6
   - python=3.8
+  - pyxlsb=1.0.8  
```
```suggestion
  - pyxlsb>=1.0
```

aleeciu replied:
done

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,322 @@
+        selected_subsec : str or list
+            Positions of the selected sectors. These positions can be either
+            defined by the user by passing a list of values, or by using built-in
+            sectors' aggregations passing a string with possible values being
```
```suggestion
            sectors' aggregations for the WIOD dat passing a string with possible values being
```

aleeciu replied:
accepted

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,322 @@
+            imp_year_set = np.array(list(imp.calc_impact_year_set(imp).values()))
+
+            # Total production of country
+            cntry_iso3 = countries_by_numeric.get(str(cntry)).alpha3
+            idx_country = np.where(self.countries_iso3 == cntry_iso3)[0]
```
Simplifying this would make it possible to use `admin1` region if one can supply the corresponding WIOT table.

aleeciu replied:
You need it as ROW is not an ISO code.

chahank replied:
I do not understand why this would be important? The only thing is you need to map the `reg_id` attribute to the WIOT attributes.
The `self.countries_iso3` is directly obtained from the WIOD table in the canonical use, and thus contains `ROW`. The cntry is obtained from the `red_ig` attribute of the exposures. Hence, it is the most natural thing to require the user to set the `reg_id` to `ROW` for the countries that belong to it right?

aleeciu replied:
There are about 40 countries in WIOD + ROW. The user should in principle be able to run a global analysis. Do we really want him to preprocess the exposures and 1) tranform the reg_id into iso3 codes and 2) filter the transformed reg_id to assess which ones are ROW and which ones can instead be found into WIOD? I do not think this is what we need.

chahank replied:
Good point about user friendliness! At the same time, hiding data structure assumptions deep in the code without documentation is not a good idea. Your code now assumes an exposure with country iso3 codes as `reg_id`. Not all exposures are necessarily this way, also not global ones. Worse, this is not even documented in the docstring! 

Hence, one possibility would be to write a separate helper function to convert an exposure with your assumed country iso3 code to the WIOD naming. 

The way it is now is not good because it assumes something about the exposures that is a) not necessary b) not documented c) restricts the code without reason.

aleeciu replied:
done in https://github.com/CLIMADA-project/climada_python/pull/180/commits/ed468290bf322d343c230549529e51c09832e9d0

chahank replied:
Amazing update!

emanuel-schmid commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,342 @@
+        self.io_data = {}
+
+    def read_wiod16(self, year=2014,
+                    table_info_pos={'range_rows':(5, 2469),
+                                    'range_cols':(4, 2468),
+                                    'col_iso3':2,'col_sectors':1}):
```
```suggestion
                    range_rows=(5, 2469),
                    range_cols=(4, 2468),
                    col_iso3=2, col_sectors=1):
```
this would imply changes in the code and docstring, of course.

chahank replied:
Another possibility is 
```
(...
table_info_post = None):
if table_info_post is None:
   table_info_pos={'range_rows':(5, 2469),
                              'range_cols':(4, 2468),
                               'col_iso3':2,
                               'col_sectors':1}
```

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,342 @@
+        self.total_aai_agg = self.total_impact.mean(axis=0)
+
+    def _map_exp_to_mriot(self, exp_regid, mriot_type):
+        """Map regions names in exposure into theiInput-output regions names."""
```
```suggestion
        """Map regions names in exposure into their Input-output regions names."""
```

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,342 @@
+        """Map regions names in exposure into theiInput-output regions names."""
+
+        if mriot_type == 'WIOD':
+            mriot_reg_name = countries_by_numeric.get(str(exp_regid)).alpha3
```
This line only works if `exp_regid` is in the format readable by `countries_by_numeric`.

aleeciu replied:
Yes, but that is what is supposed to be for a WIOD type of table, which has countries only. So one uses exposure at the country level.

chahank replied:
Yes, but I can use different standards for naming the countries. There is not only ISO3 strings. Or I could use a different country naming scheme.

aleeciu replied:
Is there such a funcionality in CLIMADA exposures, i.e., changing reg_id type? If yes, I am missing that and I will add the comment. If not, I see it as going back to WIOD-oriented (un-needed) recommendations

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,342 @@
+        hazard : Hazard
+            Hazard object for impact calculation.
+        exposure : Exposures
+            Exposures object for impact calculation.
```
Add that when using the WIOD data, the `reg_id` of the exposure must conform to ... format.

aleeciu replied:
But that is trivial, isn't it? If you use a table with country-wide exposures, then you use countries' exposures; and in CLIMADA countries exposure are supported by countries_by_numeric (see also other comment).

chahank replied:
Countries exposure are not necessarily supported by `countries_by_numeric`. Only exposures that use the ISO3 standard do so. This is no big deal. It just must be documented.

chahank commented:
Thanks for the hard work on this module! I would just add the information about the `exposure.gdf.reg_id` and solve the point about the default dict.

aleeciu commented:
> Thanks for the hard work on this module! I would just add the information about the `exposure.gdf.reg_id` and solve the point about the default dict.

Both done now in commit https://github.com/CLIMADA-project/climada_python/commit/d8d7847fa3cb45333237ec9b1e0a2f9ebce749dd. Thanks both @chahank and @emanuel-schmid for the good tips; I learned a couple of things.

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,344 @@
+        if table_info_pos is None:
+            table_info_pos = {'range_rows':(5, 2469),
+                              'range_cols':(4, 2468),
+                              'col_iso3':2,'col_sectors':1}
```
Cosmetics :D
```suggestion
                              'col_iso3':2,
                              'col_sectors':1}
```

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,344 @@
+        hazard : Hazard
+            Hazard object for impact calculation.
+        exposure : Exposures
+            Exposures object for impact calculation. For WIOD tables, countries
+            names need to follow ISO3 codes.
```
```suggestion
            Exposures object for impact calculation. For WIOD tables, exposure.region_id
            must be country names following ISO3 codes.
```

chahank commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,344 @@
+        self.total_aai_agg = self.total_impact.mean(axis=0)
+
+    def _map_exp_to_mriot(self, exp_regid, mriot_type):
+        """Map regions names in exposure into Input-output regions names."""
```
```suggestion
        """
        Map regions names in exposure into Input-output regions names.
        exp_regid must be conform to ISO3 country code format.
        """
```

emanuel-schmid commented on climada/engine/supplychain.py:
```diff
@@ -0,0 +1,345 @@
+        year : int
+            Year of WIOD table to use. Valid years go from 2000 to 2014.
+            Default year is 2014.
+        table_info_pos: dict
+            Dictionary with info about how to properly read the WIOD table.
+            Default is None. Required keys are:
+                range_rows : tuple
+                    initial and end positions of data along rows.
+                range_cols : tuple
+                    initial and end positions of data along columns.
+                col_iso3 : int
+                    column with countries names in ISO3 codes.
+                col_sectors : int
+                    column with sector names.
```
Strictly speaking, a dictionary with required keys as an argument is pretty bad style. 😉  🙊

aleeciu replied:
I am not to attached to any of the two options: I just didn't want too many kwargs; but can do either...

chahank replied:
I actually also tought that dictionaries are a good way to reduce the number of kwargs in certain cases :D

emanuel-schmid replied:
I did actually not. 🤔 The parameters of a method should be laid out plain in the open and not hidden in a dictionary. Reducing argument number with a dictionary is basically cheating and makes the situation worse not better. No complexity is reduced, just moved out of sight. Leaving the developer using the method with the burden to figure out the ''real" signature. How? Docstring? Often not in sync. Code? Often hard to read.\
High number of arguments is a hint that something in the design is suboptimal. That's the first place to look into. 
In the case at hand, all of the `table_info` arguments _could_ be derived from the Excel file itself. That would be the highest potential when it comes to reducing argument numbers.\
However, deriving these figures takes some effort and there are situations where the combination of multiple arguments into a single one makes sense. But then, _please_ use a data class instead of a dictionary. Because a dictionary is _per se_ unrestricted and unknown and a data class is not.\
In the present situation the definition of a data class is probably over the top. There are "only" 5 arguments after all.

chahank replied:
Then I probably misunderstood you at some point. Thanks for the detailed reply!

aleeciu replied:
many thanks @emanuel-schmid - dictionary removed in commit https://github.com/CLIMADA-project/climada_python/pull/180/commits/eedc8ae0ece66cb2c38d5398b04849ec9da56c82.

chahank commented:
Thanks @aleeciu , This module is now in a very good shape! The only remaining point I think is about the tests and the file directory. This requires input from @emanuel-schmid .

chahank commented:
If not yet done, could you please do a pylint check before merging? Thanks!

aleeciu commented:
> If not yet done, could you please do a pylint check before merging? Thanks!

Done already. I just don't know whether I can merge it or not :-)
